sympy/matrices/matrices.py

```diff
@@ -3957,25 +3958,28 @@ def row_insert(self, pos, mti):
         row
         col_insert
         """
-        if pos == 0:
-            return mti.col_join(self)
-        elif pos < 0:
-            pos = self.rows + pos
+        from sympy.matrices.sparse import MutableSparseMatrix, SparseMatrix
+        i = pos
         if pos < 0:
-            pos = 0
-        elif pos > self.rows:
-            pos = self.rows
-
-        if self.cols != mti.cols:
+            pos = self.rows + pos
+        if pos < -1 or pos > self.rows:
+            raise IndexError("Index out of bounds: 'pos = %s', valid -%s <= pos <= %s" 
+                % (i, self.rows + 1, self.rows))
+        elif self.cols != mti.cols:
             raise ShapeError(
-                "`self` and `mti` must have the same number of columns.")
-
-        newmat = self.zeros(self.rows + mti.rows, self.cols)
-        i, j = pos, pos + mti.rows
-        newmat[:i, :] = self[:i, :]
-        newmat[i: j, :] = mti
-        newmat[j:, :] = self[i:, :]
-        return newmat
+                    "`self` and `mti` must have the same number of columns.")
+        else:
+            if pos == -1:
+                pos = 0
+            k1 = self.rows
+            self.rows += mti.rows
+            if not isinstance(self, MutableSparseMatrix):
+                self._mat[k1*mti.cols:k1*mti.cols] = [0 for i in range(0, mti.rows*self.cols)]
```

Comment:
Since the `SparseMatrix` automatically fills in the `0's` in it after row-size increment. So i have done here separately for non-sparse matrix, since of the reason "why should i apply the same operation in SparseMatrix when i can get it's `0's` just by row-size increment". Same for col_insert